microdnf/plugins.cpp

```diff
@@ -0,0 +1,156 @@
+};
+
+PluginLibrary::PluginLibrary(const std::string & library_path) : library(library_path) {
+    get_api_version = reinterpret_cast<TGetApiVersionFunc>(library.get_address("microdnf_plugin_get_api_version"));
+    get_name = reinterpret_cast<TGetNameFunc>(library.get_address("microdnf_plugin_get_name"));
+    get_version = reinterpret_cast<TGetVersionFunc>(library.get_address("microdnf_plugin_get_version"));
+    new_instance = reinterpret_cast<TNewInstanceFunc>(library.get_address("microdnf_plugin_new_instance"));
+    delete_instance = reinterpret_cast<TDeleteInstanceFunc>(library.get_address("microdnf_plugin_delete_instance"));
+    iplugin_instance = new_instance(PLUGIN_API_VERSION);
+}
+
+PluginLibrary::~PluginLibrary() {
+    finish();
+    delete_instance(iplugin_instance);
+    iplugin_instance = nullptr;
+}
+
+Plugins::~Plugins() {
+    finish();
+    for (auto plugin = plugins.rbegin(), stop = plugins.rend(); plugin != stop; ++plugin) {
+        (*plugin).reset();
+    }
+}
+
+void Plugins::register_plugin(std::unique_ptr<Plugin> && plugin) {
+    auto logger = context->base.get_logger();
+    auto * iplugin = plugin->get_iplugin();
+    plugins.emplace_back(std::move(plugin));
+    auto name = iplugin->get_name();
+    auto version = iplugin->get_version();
+    logger->debug(fmt::format(
+        "Added plugin name=\"{}\", version=\"{}.{}.{}\"", name, version.major, version.minor, version.micro));
+}
+
+void Plugins::load_plugin(const std::string & file_path) {
+    auto logger = context->base.get_logger();
+    logger->debug(fmt::format("Loading plugin file=\"{}\"", file_path));
+    auto plugin = std::make_unique<PluginLibrary>(file_path);
+    auto * iplugin = plugin->get_iplugin();
+    plugins.emplace_back(std::move(plugin));
+    auto name = iplugin->get_name();
+    auto version = iplugin->get_version();
+    logger->debug(fmt::format(
+        "Loaded plugin name=\"{}\", version=\"{}.{}.{}\"", name, version.major, version.minor, version.micro));
+}
+
+void Plugins::load_plugins(const std::string & dir_path) {
+    auto logger = context->base.get_logger();
+    if (dir_path.empty())
+        throw std::runtime_error(_("Plugins::loadPlugins() dirPath cannot be empty"));
+
+    std::vector<std::filesystem::path> lib_names;
+    for (auto & p : std::filesystem::directory_iterator(dir_path)) {
+        if ((p.is_regular_file() || p.is_symlink()) && p.path().extension() == ".so") {
+            lib_names.emplace_back(p.path());
+        }
+    }
+    std::sort(lib_names.begin(), lib_names.end());
+
+    bool cant_load{false};
+    for (auto & p : lib_names) {
+        try {
+            load_plugin(p);
+        } catch (const std::exception & ex) {
+            std::string msg = fmt::format("Cannot load plugin \"{}\": {}", p.string(), ex.what());
+            logger->error(msg);
+            cant_load = true;
+        }
+    }
+
+    if (cant_load) {
+        throw std::runtime_error(M_("Cannot load plugins"));
```

Comment:
Exception class; also what if we at least made a list of the plugins that failed to load to include in the error message? We'd then print the error to stderr for the user and they'd know without digging in the log what plugins are causing issues (granted a look into the log is usually needed to troubleshoot anyway).

Reply:
I will add the paths of the libraries that failed to load to the error message.